ruby/ext/google/protobuf_c/message.c

```diff
@@ -905,6 +905,33 @@ static VALUE Message_freeze(VALUE _self) {
   return _self;
 }
 
+/*
+ * call-seq:
+ *     Message.deep_freeze => self
+ *
+ * Deeep freezes the message object recursively.
+ */
+static VALUE Message_deep_freeze(VALUE _self) {
+  Message* self = ruby_to_Message(_self);
+  if (!RB_OBJ_FROZEN(_self)) {
+    Message_freeze(_self);
+
+    int n = upb_MessageDef_FieldCount(self->msgdef);
+    for (int i = 0; i < n; i++) {
+      const upb_FieldDef* f = upb_MessageDef_Field(self->msgdef, i);
+      VALUE field = Message_getfield(_self, f);
+
+      if (upb_FieldDef_IsMap(f) || upb_FieldDef_IsRepeated(f)) {
+        rb_funcall(field, rb_intern("freeze"), 0);
```

Comment:
Repeated fields and maps can contain messages, so we will need them to support `deep_freeze` also.